src/store/ducks/opensrp/hierarchies/utils.tsx

```diff
@@ -202,82 +201,79 @@ export function computeParentNodesSelection(
       metaDataByJurisdiction
     );
     if (allChildrenAreSelected) {
-      metaDataByJurisdictionId = {
-        ...metaDataByJurisdictionId,
-        ...createSingleMetaData(parentNode.model.id, actionBy, true),
-      };
+      metaDataByJurisdictionId = Object.assign(
+        metaDataByJurisdictionId,
+        createSingleMetaData(parentNode.model.id, actionBy, true)
+      );
     } else {
       break;
     }
   }
   return metaDataByJurisdictionId;
 }
 
+/** compute Selected nodes from the tree
+ * @param tree - the whole tree or undefined
+ * @param leafNodesOnly - whether to include the leaf nodes only
+ * @param metaByJurisdiction - MetaData object
+ */
+export const computeSelectedNodes = (
+  tree: TreeNode | undefined,
+  leafNodesOnly: boolean | undefined,
+  metaByJurisdiction: Dictionary<Meta>
+) => {
+  const nodesList: TreeNode[] = [];
+  if (!tree) {
+    return [];
+  }
+  tree.walk(node => {
+    const shouldAddNode =
+      !!(metaByJurisdiction[node.model.id] && metaByJurisdiction[node.model.id][SELECTION_KEY]) &&
+      !(leafNodesOnly && node.hasChildren());
+    if (shouldAddNode) {
+      nodesList.push(node);
+    }
+    return true;
+  });
+  return nodesList;
+};
+
 /** any callback that will take a node and return whether we should select the node
+ * @param tree - the tree
  * @param callback - callback is given each node in a walk and decides whether that node should be selected by returning true or false
+ * @param actionBy - selected by
  */
 export const autoSelectNodesAndCascade = (
   tree: TreeNode,
   callback: (node: TreeNode) => boolean,
-  actionBy: string = SELECTION_REASON.USER_CHANGE,
-  thisTreeMetaData: Dictionary<Meta>
+  actionBy: string = SELECTION_REASON.USER_CHANGE
 ) => {
   let metaByJurisdictionId: Dictionary<Meta> = {};
-  // deselect all
-  const { metaByJurisdiction } = setSelectOnNode(tree, tree.model.id, actionBy, false, true, true);
-  const parentNodes: TreeNode[] = [];
+
   tree.walk(node => {
     if (callback(node)) {
       node.walk(nd => {
-        metaByJurisdictionId = {
-          ...createSingleMetaData(nd.model.id, actionBy, true),
-          ...metaByJurisdictionId,
-        };
+        metaByJurisdictionId = Object.assign(
+          metaByJurisdictionId,
+          createSingleMetaData(nd.model.id, actionBy, true)
+        );
         // removing this return cause a type error, that's its sole purpose
         return true;
       });
-      parentNodes.push(node);
     }
     return true;
```

Comment:
Is it possible to refactor this so that we only have one `return true;` statement?

Reply:
AFAIK, am afraid not.